File: src/rbi/apply/resolve.py
"""The as-of resolver (CLAUDE.md §5). Entity + validity filter, then the answer.

This is the query that defines the product. It never guesses: if nothing is in
force on `as_of`, it abstains with a status that says *why* (not yet in force /
no longer in force / no provision) and lists the versions it considered.
"""
from __future__ import annotations

from datetime import date

from .models import ClauseVersion, Resolution


def resolve(
    versions: list[ClauseVersion],
    *,
    md_family: str,
    entity_type_code: str,
    clause_number: str,
    as_of: date,
) -> Resolution:
    # Entity + clause filter FIRST — semantics never cross entity types.
    candidates = [
        v
        for v in versions
        if v.md_family == md_family
        and v.entity_type_code == entity_type_code
        and v.clause_number == clause_number
    ]
    base = dict(
        md_family=md_family,
        entity_type_code=entity_type_code,
        clause_number=clause_number,
        as_of=as_of,
        candidates=sorted(candidates, key=lambda v: v.valid_from),
    )

    if not candidates:
        return Resolution(
            status="no_provision",
            note=f"clause {clause_number} does not exist for {entity_type_code}",
            **base,
        )

    for v in candidates:
        if v.valid_from <= as_of and (v.valid_to is None or v.valid_to > as_of):
            return Resolution(
                status="in_force",
                text=v.text,
                valid_from=v.valid_from,
                valid_to=v.valid_to,
                **base,
            )

    future = [v for v in candidates if v.valid_from > as_of]
    if future:
        soonest = min(future, key=lambda v: v.valid_from)
        return Resolution(
            status="not_yet_in_force",
            effective_date=soonest.valid_from,
            note=(
                f"clause {clause_number} was issued but comes into force "
                f"{soonest.valid_from}; on {as_of} the prior text (if any) applies"
            ),
            **base,
        )

    latest = max(candidates, key=lambda v: v.valid_from)
    return Resolution(
        status="no_longer_in_force",
        valid_to=latest.valid_to,
        note=f"clause {clause_number} was closed on {latest.valid_to}",
        **base,
    )

File: api/src/resolve.js
// Port of rbi.apply.resolve (Python) — the as-of resolver, verified by its own tests.
// Kept behaviourally identical so Python and JS agree on the answer.
//
// Dates are ISO strings ('YYYY-MM-DD'); string comparison orders them correctly,
// so we avoid Date objects and timezone surprises entirely.

/** @typedef {Object} ClauseVersion
 *  @property {string} mdFamily
 *  @property {string} entityCode
 *  @property {string} clauseNumber
 *  @property {string} text
 *  @property {string} validFrom  ISO date
 *  @property {string|null} validTo ISO date or null (still in force)
 */

export const ANSWERABLE = new Set(['in_force', 'equivalence', 'cascade']);

/**
 * Resolve what a clause says for an entity on a date. Never guesses: if nothing is
 * in force it abstains with a status explaining why, plus the candidates considered.
 */
export function resolve(versions, { mdFamily, entityCode, clauseNumber, asOf }) {
  // Entity + clause filter FIRST — semantics never cross entity types.
  const candidates = versions
    .filter(
      (v) =>
        v.mdFamily === mdFamily &&
        v.entityCode === entityCode &&
        v.clauseNumber === clauseNumber,
    )
    .sort((a, b) => (a.validFrom < b.validFrom ? -1 : a.validFrom > b.validFrom ? 1 : 0));

  const base = { mdFamily, entityCode, clauseNumber, asOf, candidates };

  if (candidates.length === 0) {
    return {
      ...base,
      status: 'no_provision',
      text: null,
      validFrom: null,
      validTo: null,
      effectiveDate: null,
      note: `clause ${clauseNumber} does not exist for ${entityCode}`,
    };
  }

  for (const v of candidates) {
    if (v.validFrom <= asOf && (v.validTo === null || v.validTo > asOf)) {
      return {
        ...base,
        status: 'in_force',
        text: v.text,
        validFrom: v.validFrom,
        validTo: v.validTo,
        effectiveDate: null,
        note: null,
        sourceRef: v.sourceRef ?? null,
        issuedDate: v.issuedDate ?? null,
      };
    }
  }

  const future = candidates.filter((v) => v.validFrom > asOf);
  if (future.length > 0) {
    const soonest = future.reduce((a, b) => (a.validFrom < b.validFrom ? a : b));
    return {
      ...base,
      status: 'not_yet_in_force',
      text: null,
      validFrom: null,
      validTo: null,
      effectiveDate: soonest.validFrom,
      note:
        `clause ${clauseNumber} was issued but comes into force ${soonest.validFrom}; ` +
        `on ${asOf} the prior text (if any) applies`,
    };
  }

  const latest = candidates.reduce((a, b) => (a.validFrom > b.validFrom ? a : b));
  return {
    ...base,
    status: 'no_longer_in_force',
    text: null,
    validFrom: null,
    validTo: latest.validTo,
    effectiveDate: null,
    note: `clause ${clauseNumber} was closed on ${latest.validTo}`,
  };
}
